### PR TITLE
Document numpy traversals

### LINKED ISSUE #1788: Document traversal arrays and illustrate usage
#1704 added numpy array versions of the basic tree traversal orders. Document these and provide an example of how they might be used. It would be neat if we could illustrate with some phylo algorithm that can be implemented directly with (non-obscure) numpy calls.

### PR DESCRIPTION
**Pull request by benjeffery**
Fixes #1788 

Added for follow up work https://github.com/tskit-dev/tutorials/issues/150

### CODE REVIEW

**Comment by hyanwong** [comment]
Yay, I was just about to do this @benjeffery but you beat me to it!

A couple of things:

1) you might be able to save some repetition by nuking some of the text in the descriptions of orders in the `Tree.nodes()` docketing, and simply pointing to these new methods, e.g. 

```
... The available orders are:

        - 'preorder': Standard preorder traversal, see :meth:`Tree.preorder`.
        - ...
```

2) The docs don't say what we do if there are multiple roots in a tree. I think we do everything from one root, then everything from the next root, etc, but don't guarantee the order in which we iterate over the roots, right? This could also do with clarifying for e.g. `inorder` and `levelorder` in the `Tree.nodes()` docketing, but that's a different issue I guess.

File: python/tskit/trees.py
Comment (hyanwong) [inline python/tskit/trees.py:2150]: Should we say "increasing time (youngest first)" so it's clear what direction time goes in?

**Review by hyanwong — APPROVED** [review]

File: python/tskit/trees.py
Comment (jeromekelleher) [inline python/tskit/trees.py:2127]: This isn't quite right. If we start from the virtual root we return the virtual root in the arrays, but this isn't what we usually want. It's more like "if specified, return all nodes in the subtree rooted at u (including u) in traversal order".

File: python/tskit/trees.py
Comment (jeromekelleher) [inline python/tskit/trees.py:2121]: This isn't quite right I think. How about 

> Returns a number array of node ids in `preorder <https://en.wikipedia.org/wiki/Tree_traversal#Pre-order_(NLR)>`. If the node ``u`` the traversal is rooted at this node (and it will be the first element in the returned array). Otherwise, all nodes reachable from the tree roots will be returned.  See :ref:`tutorials:sec_analysing_trees_traversals` for examples.

I don't think need to describe what "preorder" is, the wikipedia link is fine. I think talking about "yielding" here is more confusing, since this isn't a generator.

**Review by jeromekelleher — COMMENTED** [review]
Looks good, needs some tweaks as the semantics aren't quite right.

File: python/tskit/trees.py
Comment (benjeffery) [reply python/tskit/trees.py]: Done

File: python/tskit/trees.py
Comment (benjeffery) [reply python/tskit/trees.py]: Fixed

File: python/tskit/trees.py
Comment (benjeffery) [reply python/tskit/trees.py]: Fixed

**Comment by benjeffery** [comment]
Should be good to go now

File: python/tskit/trees.py
Comment (hyanwong) [inline python/tskit/trees.py:2141]: Does this need backticks:

```
:ref:`tutorials:sec_analysing_trees_traversals`
```

File: python/tskit/trees.py
Comment (hyanwong) [inline python/tskit/trees.py:2125]: Does this need backticks:

```
:ref:`tutorials:sec_analysing_trees_traversals`
```

**Review by hyanwong — COMMENTED** [review]
Trivial nitpicks

File: python/tskit/trees.py
Comment (hyanwong) [inline python/tskit/trees.py:2139]: "last element"?

**Comment by hyanwong** [comment]
> Should be good to go now

Great. It might be good to add something about the order of traversal if there are multiple roots. Something like:

```
Returns a 1D numpy array of node ids in preorder
<https://en.wikipedia.org/wiki/Tree_traversal#Pre-order_(NLR)>. If the node u
is specified the traversal is rooted at this node (and it will be the first
element in the returned array). Otherwise, return an array in which the node
in preorder reachable from each tree root have been concatenated together.
```

but I can add that as a further PR if you like. It would be good to merge this, then I can push my doc PR properly.

File: python/tskit/trees.py
Comment (hyanwong) [inline python/tskit/trees.py:2121]: number -> numpy?

File: python/tskit/trees.py
Comment (hyanwong) [inline python/tskit/trees.py:2153]: number -> numpy

File: python/tskit/trees.py
Comment (hyanwong) [inline python/tskit/trees.py:2137]: number -> numpy?

**Review by hyanwong — COMMENTED** [review]
More nitpicks, sorry!

File: python/tskit/trees.py
Comment (benjeffery) [reply python/tskit/trees.py]: Done

File: python/tskit/trees.py
Comment (benjeffery) [reply python/tskit/trees.py]: Done

**Comment by benjeffery** [comment]
Thanks @hyanwong should be good now

File: python/tskit/trees.py
Comment (jeromekelleher) [inline python/tskit/trees.py:2104]: Some collatoral damage here on s/number/numpy/g!

File: python/tskit/trees.py
Comment (jeromekelleher) [reply python/tskit/trees.py:2125]: Yes it does

**Comment by jeromekelleher** [comment]
We should discuss the semantics of the virtual root too - but, possibly simpler to merge this once the cross references have been fixed so that @hyanwong can merge in his version?

File: python/tskit/trees.py
Comment (benjeffery) [reply python/tskit/trees.py:2104]: Oh boy. :man_facepalming:

**Comment by benjeffery** [comment]
Fixed

**Comment by benjeffery** [comment]
@mergifyio rebase

File: python/tskit/trees.py
Comment (benjeffery) [reply python/tskit/trees.py]: Done

File: python/tskit/trees.py
Comment (benjeffery) [reply python/tskit/trees.py]: Done

**Comment by benjeffery** [comment]
Ok, hopefully that does it. FML.

**Comment by hyanwong** [comment]
Do you want me to merge stuff into this before it's merged into `main`, or simply make a follow-up PR? I've never done the former, but happy to try.

**Comment by jeromekelleher** [comment]
Let's do a follow-up @hyanwong

**Review by jeromekelleher — APPROVED** [review]

**Pull request merged**
